From 890f21aab6090639c9af97889bda099857f59f60 Mon Sep 17 00:00:00 2001
From: Connor Abbott <cwabbott0@gmail.com>
Date: Thu, 21 May 2026 12:10:41 -0400
Subject: [PATCH] ir3/opt_preamble: Consider speculatability correctly

Now that producers set it, stop setting it on everything. Base
descriptor speculatability on the information in ir3_bindless_resource
now that it is plumbed through.

Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/42159>
(cherry picked from mesa main commit 890f21aab6090639c9af97889bda099857f59f60)
---
 src/compiler/nir/nir_intrinsics.py       |   6 +-
 src/freedreno/ir3/ir3_nir.h              |  12 +++
 src/freedreno/ir3/ir3_nir_opt_preamble.c | 119 +++++++++--------------
 3 files changed, 63 insertions(+), 74 deletions(-)

diff --git a/src/compiler/nir/nir_intrinsics.py b/src/compiler/nir/nir_intrinsics.py
index 04b6abc3527..35af592b184 100644
--- a/src/compiler/nir/nir_intrinsics.py
+++ b/src/compiler/nir/nir_intrinsics.py
@@ -1737,9 +1737,9 @@ intrinsic("shuffle_down_uniform_ir3", src_comp=[0, 1], dest_comp=0,
           bit_sizes=src0, flags=[CAN_ELIMINATE])
 
 # IR3-specific intrinsics for prefetching descriptors in preambles.
-intrinsic("prefetch_sam_ir3", [1, 1], flags=[CAN_REORDER])
-intrinsic("prefetch_tex_ir3", [1], flags=[CAN_REORDER])
-intrinsic("prefetch_ubo_ir3", [1], flags=[CAN_REORDER])
+intrinsic("prefetch_sam_ir3", [1, 1], indices=[ACCESS], flags=[CAN_REORDER])
+intrinsic("prefetch_tex_ir3", [1], indices=[ACCESS], flags=[CAN_REORDER])
+intrinsic("prefetch_ubo_ir3", [1], indices=[ACCESS], flags=[CAN_REORDER])
 
 intrinsic("resbase_ir3", src_comp=[1], dest_comp=2, flags=[CAN_ELIMINATE, CAN_REORDER])
 
diff --git a/src/freedreno/ir3/ir3_nir.h b/src/freedreno/ir3/ir3_nir.h
index 9f6c82216bd..8065121b059 100644
--- a/src/freedreno/ir3/ir3_nir.h
+++ b/src/freedreno/ir3/ir3_nir.h
@@ -214,6 +214,18 @@ nir_io_offset ir3_nir_get_global_offset(nir_builder *b,
                                         struct ir3_compiler *compiler,
                                         nir_def *offset, unsigned offset_shift);
 
+/* Returns true if an intrinsic reading memory can be safely moved into the
+ * preamble. It's assumed the intrinsic is already reorderable. Either it must
+ * be always executed, or it must be safe to execute even if the original shader
+ * never executed it (i.e. it must be speculatable).
+ */
+static inline bool
+ir3_nir_is_prefetchable(nir_intrinsic_instr *intr)
+{
+   return intr->instr.block->cf_node.parent->type == nir_cf_node_function ||
+      (nir_intrinsic_access(intr) & ACCESS_CAN_SPECULATE);
+}
+
 ENDC;
 
 #endif /* IR3_NIR_H_ */
diff --git a/src/freedreno/ir3/ir3_nir_opt_preamble.c b/src/freedreno/ir3/ir3_nir_opt_preamble.c
index 43dd7e07b3e..1e90d1dd3b9 100644
--- a/src/freedreno/ir3/ir3_nir_opt_preamble.c
+++ b/src/freedreno/ir3/ir3_nir_opt_preamble.c
@@ -277,44 +277,6 @@ avoid_instr(const nir_instr *instr, const void *data)
    return intrin->intrinsic == nir_intrinsic_bindless_resource_ir3;
 }
 
-static bool
-set_speculate(nir_builder *b, nir_instr *instr, UNUSED void *_)
-{
-   if (instr->type == nir_instr_type_tex) {
-      nir_instr_as_tex(instr)->can_speculate = true;
-      return true;
-   }
-
-   if (instr->type != nir_instr_type_intrinsic)
-      return false;
-
-   nir_intrinsic_instr *intr = nir_instr_as_intrinsic(instr);
-
-   switch (intr->intrinsic) {
-   /* These instructions go through bounds-checked hardware descriptors so
-    * should be safe to speculate.
-    *
-    * TODO: This isn't necessarily true in Vulkan, where descriptors don't need
-    * to be filled out and bindless descriptor offsets aren't bounds checked.
-    * We may need to plumb this information through from turnip for correctness
-    * to avoid regressing freedreno codegen.
-    */
-   case nir_intrinsic_load_ubo:
-   case nir_intrinsic_load_ubo_vec4:
-   case nir_intrinsic_image_load:
-   case nir_intrinsic_image_samples_identical:
-   case nir_intrinsic_bindless_image_load:
-   case nir_intrinsic_load_ssbo:
-   case nir_intrinsic_load_ssbo_ir3:
-      nir_intrinsic_set_access(intr, nir_intrinsic_access(intr) |
-                                     ACCESS_CAN_SPECULATE);
-      return true;
-
-   default:
-      return false;
-   }
-}
-
 bool
 ir3_nir_opt_preamble(nir_shader *nir, struct ir3_shader_variant *v)
 {
@@ -332,9 +294,6 @@ ir3_nir_opt_preamble(nir_shader *nir, struct ir3_shader_variant *v)
    if (max_size == 0)
       return false;
 
-   bool progress = nir_shader_instructions_pass(nir, set_speculate,
-                                                nir_metadata_control_flow, NULL);
-
    nir_opt_preamble_options options = {
       .drawid_uniform = true,
       .subgroup_size_uniform = true,
@@ -348,7 +307,7 @@ ir3_nir_opt_preamble(nir_shader *nir, struct ir3_shader_variant *v)
    };
 
    unsigned size = 0;
-   progress |= nir_opt_preamble(nir, &options, &size);
+   bool progress = nir_opt_preamble(nir, &options, &size);
 
    if (!v->binning_pass) {
       uint32_t preamble_size_vec4 =
@@ -375,16 +334,15 @@ ir3_def_is_rematerializable_for_preamble(nir_def *def,
       nir_intrinsic_instr *intrin = nir_def_as_intrinsic(def);
       switch (intrin->intrinsic) {
       case nir_intrinsic_load_ubo:
-         return ir3_def_is_rematerializable_for_preamble(intrin->src[0].ssa,
-                                                         preamble_defs) &&
-            ir3_def_is_rematerializable_for_preamble(intrin->src[1].ssa,
+         return ir3_nir_is_prefetchable(intrin) &&
+            ir3_def_is_rematerializable_for_preamble(intrin->src[0].ssa,
                                                      preamble_defs) &&
-            (nir_def_block(def)->cf_node.parent->type ==
-             nir_cf_node_function ||
-             (nir_intrinsic_access(intrin) & ACCESS_CAN_SPECULATE));
+            ir3_def_is_rematerializable_for_preamble(intrin->src[1].ssa,
+                                                     preamble_defs);
       case nir_intrinsic_bindless_resource_ir3:
-         return ir3_def_is_rematerializable_for_preamble(intrin->src[0].ssa,
-                                                         preamble_defs);
+         return ir3_nir_is_prefetchable(intrin) &&
+            ir3_def_is_rematerializable_for_preamble(intrin->src[0].ssa,
+                                                     preamble_defs);
       case nir_intrinsic_load_preamble:
          return !!preamble_defs;
       default:
@@ -605,6 +563,33 @@ get_descriptors(nir_instr *instr, nir_def **descs)
    }
 }
 
+static bool
+is_descriptor_prefetch_speculatable(nir_def *desc)
+{
+   nir_instr *instr = nir_def_instr(desc);
+
+   /* Non-bindless descriptors are always speculatable */
+   if (instr->type != nir_instr_type_intrinsic)
+      return true;
+
+   nir_intrinsic_instr *bindless = nir_instr_as_intrinsic(nir_def_instr(desc));
+
+   if (bindless->intrinsic != nir_intrinsic_bindless_resource_ir3)
+      return true;
+
+   return nir_intrinsic_access(bindless) & ACCESS_CAN_SPECULATE;
+}
+
+static bool
+is_descriptor_prefetchable(nir_def *desc)
+{
+   nir_instr *instr = nir_def_instr(desc);
+
+   return instr->block->cf_node.parent->type == nir_cf_node_function ||
+      is_descriptor_prefetch_speculatable(desc);
+}
+
+
 #define MAX_PREFETCHES 32
 
 struct prefetches {
@@ -636,10 +621,12 @@ struct prefetch_state {
 
 static bool
 emit_descriptor_prefetch(nir_builder *b, nir_instr *instr, nir_def **descs,
-                         struct prefetch_state *state)
+                         struct prefetch_state *state, bool can_speculate)
 {
    nir_block *insert_block = nir_def_block(descs[0]);
 
+   enum gl_access_qualifier access = can_speculate ? ACCESS_CAN_SPECULATE : 0;
+
    if (descs[1]) {
       insert_block = find_insert_block_for_defs(descs, 2);
 
@@ -688,13 +675,13 @@ emit_descriptor_prefetch(nir_builder *b, nir_instr *instr, nir_def **descs,
          if (!sampler_already_prefetched)
             add_prefetch(&state->sampler, descs[1]);
 
-         nir_prefetch_sam_ir3(b, descs[0], descs[1]);
+         nir_prefetch_sam_ir3(b, descs[0], descs[1], .access = access);
       } else {
          if (tex_already_prefetched)
             return false;
 
          add_prefetch(&state->tex, descs[0]);
-         nir_prefetch_tex_ir3(b, descs[0]);
+         nir_prefetch_tex_ir3(b, descs[0], .access = access);
       }
    } else {
       assert(instr->type == nir_instr_type_intrinsic);
@@ -709,9 +696,9 @@ emit_descriptor_prefetch(nir_builder *b, nir_instr *instr, nir_def **descs,
 
       nir_intrinsic_instr *intrin = nir_instr_as_intrinsic(instr);
       if (intrin->intrinsic == nir_intrinsic_load_ubo)
-         nir_prefetch_ubo_ir3(b, descs[0]);
+         nir_prefetch_ubo_ir3(b, descs[0], .access = access);
       else
-         nir_prefetch_tex_ir3(b, descs[0]);
+         nir_prefetch_tex_ir3(b, descs[0], .access = access);
    }
 
    return true;
@@ -785,21 +772,6 @@ ir3_nir_opt_prefetch_descriptors(nir_shader *nir, struct ir3_shader_variant *v)
                should_prefetch_descriptor(descs[1])))
             continue;
 
-         /* The instruction itself must be hoistable.
-          * TODO: If the descriptor is statically referenced and in-bounds, then
-          * we should be able to hoist the descriptor load even if the
-          * descriptor contents aren't guaranteed. This would require more
-          * plumbing.
-          * TODO: Textures. This is broken in nir_opt_preamble at the moment and
-          * handling them would also require more plumbing.
-          */
-         if (instr->type == nir_instr_type_intrinsic &&
-             nir_intrinsic_has_access(nir_instr_as_intrinsic(instr)) &&
-             !(nir_intrinsic_access(nir_instr_as_intrinsic(instr)) &
-               ACCESS_CAN_SPECULATE) &&
-             block->cf_node.parent->type != nir_cf_node_function)
-            continue;
-
          /* Each descriptor must be rematerializable */
          if (descs[0] &&
              !ir3_def_is_rematerializable_for_preamble(descs[0], preamble_defs))
@@ -808,6 +780,10 @@ ir3_nir_opt_prefetch_descriptors(nir_shader *nir, struct ir3_shader_variant *v)
              !ir3_def_is_rematerializable_for_preamble(descs[1], preamble_defs))
             continue;
 
+         bool is_speculatable =
+            (!descs[0] || is_descriptor_prefetch_speculatable(descs[0])) &&
+            (!descs[1] || is_descriptor_prefetch_speculatable(descs[1]));
+
          /* If the preamble hasn't been created then this descriptor isn't a
           * duplicate and we will definitely insert an instruction, so create
           * the preamble if it hasn't already been created.
@@ -858,7 +834,8 @@ ir3_nir_opt_prefetch_descriptors(nir_shader *nir, struct ir3_shader_variant *v)
                                                   preamble_defs);
          }
 
-         progress |= emit_descriptor_prefetch(&b, instr, preamble_descs, &state);
+         progress |= emit_descriptor_prefetch(&b, instr, preamble_descs, &state,
+                                              is_speculatable);
 
          if (state.sampler.num_prefetches == MAX_PREFETCHES &&
              state.tex.num_prefetches == MAX_PREFETCHES)
-- 
2.55.0

